firedrake/slate/slac/kernel_builder.py

```diff
@@ -379,3 +390,420 @@ def needs_mesh_layers(self):
+
+                    # TODO subdomain boundary integrals, this does the wrong thing for integrals like f*ds + g*ds(1)
+                    # "otherwise" is treated incorrectly as "everywhere"
+                    # However, this replicates an existing slate bug.
+                    if kinfo.subdomain_id != "otherwise":
+                        predicates.append("cell_facets["+str(fidx[0])+",1]=="+str(kinfo.subdomain_id))
+
+                    subscript = pym.Subscript(pym.Variable(self.local_facet_array_arg),
+                                              (pym.Sum((i[0], fidx[0]))))
+                    reads.append(SubArrayRef(i, subscript))
+                    inames.append(fidx[0].name)
+                elif integral_type in {"interior_facet_horiz_top",
+                                       "interior_facet_horiz_bottom",
+                                       "exterior_facet_top",
+                                       "exterior_facet_bottom"}:
+
+                    needs_mesh_layers = True
+                    layer = pym.Variable(self.layer_arg)
+
+                    # TODO: Variable layers
+                    nlayer = tensor.ufl_domain().layers-1
+                    which = {"interior_facet_horiz_top": str(layer)+"<"+str(nlayer-1),
+                             "interior_facet_horiz_bottom": str(layer)+">"+str(0),
+                             "exterior_facet_top": str(layer)+"=="+str(nlayer-1),
+                             "exterior_facet_bottom": str(layer)+"=="+str(0)}[integral_type]
+                    predicates = [which]
+                else:
+                    raise ValueError("Unhandled integral type {}".format(integral_type))
+
+                # Reads are the variables being fed into the subkernel
+                # later on assemby calls will be needed for the kitting instruction
+                # when merging the outer (slate) kernel with the inner (ufl) kernel
+                assembly_calls[integral_type].append(loopy.CallInstruction((output, ),
+                                                                           pym.Call(pym.Variable(kinfo.kernel.name), tuple(reads)),
+                                                                           predicates=predicates,
+                                                                           within_inames=frozenset(inames),
+                                                                           id=integral_type+"_inner_call%d" % len(assembly_calls[integral_type])))
+
+        self.assembly_calls = assembly_calls
+        self.templated_subkernels = templated_subkernels
+        self.include_dirs = list(set(include_dirs))
+        self.needs_cell_orientations = needs_cell_orientations
+        self.needs_cell_sizes = needs_cell_sizes
+        self.needs_cell_facets = needs_cell_facets
+        self.needs_mesh_layers = needs_mesh_layers
+        self.num_facets = num_facets
```

Comment:
Okay, so in general I tried to alter the code as little as possible from the previous slate compiler structure, because I thought that would be preferred. There are two main things happening in the LocalLoopyKernelBuilder that are similar to the previous builder. First we do a mapping of the Terminals (Tensor and Assembled Vector to gem). Second we setup the calls to tsfc, which fill those Terminals. The general line of thought in the setup is to gather the information that need to be fed into the TSFC loopy kernel as arguments (that's what's in kernel_data, but was previously called args in Thomas KernelBuilder), handle certain integral_types (which needs some predicates in loopy) and then save the call TSFC instruction.

I can see that this exploded a bit, so I refactored the code as it is now. After that I will think about potentially postponing some of the things that you call 

> "early" management of mappings

 in a later comment.

Reply:
I think this is a good opportunity to refactor to do things in a cleaner way. For an idea of what kind of separation I mean, here's a strawman of translation of the core gem expressions to a loopy kernel (I didn't do blocks because that took more thought than I had).

At the end, `slate2loopy` spits out a loopy kernel and a mapping from the kernel global args to slate terminal expressions. One could then generate the right sequence of calls for population of the terminal temporaries and emit the gluing kernel, I think.

```python
@singledispatch
def _slate2gem(expr, self):
    raise AssertionError


@_slate2gem.register(slate.Tensor)
@_slate2gem.register(slate.AssembledVector)
def _slate2gem_tensor(expr, self):
    shape = expr.shape
    name = f"T{len(self.var2terminal)}"
    assert expr not in self.var2terminal.values()
    var = gem.Variable(name, shape)
    self.var2terminal[var] = expr
    return var


@_slate2gem.register(slate.Block)
def _slate2gem_block(expr, self):
    raise AssertionError("Not yet done")


@_slate2gem.register(slate.Inverse)
def _slate2gem_inverse(expr, self):
    return gem.Inverse(*map(self, expr.children))


@_slate2gem.register(slate.Solve)
def _slate2gem_solve(expr, self):
    return gem.Solve(*map(self, expr.children))


@_slate2gem.register(slate.Transpose)
def _slate2gem_transpose(expr, self):
    child, = map(self, expr.children)
    indices = tuple(gem.indices(len(child.shape)))
    return gem.ComponentTensor(gem.Indexed(child, indices), tuple(indices[::-1]))


@_slate2gem.register(slate.Negative)
def _slate2gem_negative(expr, self):
    child, = map(self, expr.children)
    indices = tuple(gem.indices(len(child.shape)))
    return gem.ComponentTensor(gem.Product(gem.Literal(-1),
                                           gem.Indexed(child, indices)),
                               indices)


@_slate2gem.register(slate.Add)
def _slate2gem_add(expr, self):
    A, B = map(self, expr.children)
    indices = tuple(gem.indices(len(A.shape)))
    return gem.ComponentTensor(gem.Sum(gem.Indexed(A, indices),
                                       gem.Indexed(B, indices)),
                               indices)


@_slate2gem.register(slate.Mul)
def _slate2gem_mul(expr, self):
    A, B = map(self, expr.children)
    *i, k = tuple(gem.indices(len(A.shape)))
    _, *j = tuple(gem.indices(len(B.shape)))
    ABikj = gem.Product(gem.Indexed(A, tuple(i + [k])),
                        gem.Indexed(B, tuple([k] + j)))
    return gem.ComponentTensor(gem.IndexSum(ABikj, (k, )), tuple(i + j))


def slate2gem(expressions):
    mapper = Memoizer(_slate2gem)
    mapper.var2terminal = OrderedDict()
    return list(map(mapper, expressions)), mapper.var2terminal


def slate2loopy(expressions, parameters=None):
    params = default_parameters().copy()
    if parameters is not None:
        params.update(parameters)

    (ir, ), varmapping = slate2gem(expressions)
    retval = gem.Variable("A", ir.shape)

    args = [loopy.GlobalArg(retval.name, shape=retval.shape, dtype=params["scalar_type"])]
    argmapping = OrderedDict()
    for var, expr in varmapping.items():
        arg = loopy.GlobalArg(var.name, shape=var.shape, dtype=params["scalar_type"])
        args.append(arg)
        argmapping[arg] = expr

    indices = gem.indices(len(ir.shape))
    ir = gem.Indexed(ir, indices)
    retval = gem.Indexed(retval, indices)
    ir, = preprocess_gem([ir])
    impero_c = compile_gem([(retval, ir)], prefix_ordering=indices, remove_zeros=True)

    return generate(impero_c, args, params["precision"], params["scalar_type"]), argmapping
```

Reply:
Okay so main differences between this approach and mine are the following.

- [x] Instead of doing a pre-iteration to save translation of tensors and assembled vectors in a mapping in the kernel builder, this approach does the translation in the actual slate to gem step. -> I agree it is cleaner to do that, but as before we still have to carry this mapping to the next translation into loopy. The only difference is that now the mapper carries the mapping and not the builder.
- [ ]  Instead of saving the tree traversal in a list and do the translation item by item, you suggest to use the Memoizer, which does the translation recursively, right? -> Will this slow down the compilation process?
- [x] Inverse and Solve translations don’t need new indices anymore. -> I think, that is right. But currently I am dropping the factorisation node in the slate to gem translation of the Solve. So either we would have to modify the mapping you suggested here, or we would need to deal with it somewhere else. Currently the Solve Call always does a partial piv factorisation independent of what the user might specify.

Now the main point 
The way you wrote slate2loopy suggests that you want to not only inline the TSFC  kernel, but also the Slate kernel. You want to call Tsfc to populate some temporaries and feed this as global args to the slate loopy kernel, both inside some other wrapper kernel, I assume.
  1. The way I thought about this is, that the slate kernel *is* the wrapper kernel (which is why I call it outer loopy) and only the TSFC kernels (inner loopy) get inlined. That is also why I am already considering arguments that are actually needed for the TSFC call here https://github.com/firedrakeproject/firedrake/blob/418eaea8e307de39983cf9eb60caa8e01d266212/firedrake/slate/slac/compiler.py#L627 I think this should be moved to the step when I merge the two loopy kernels, however.
  2. Either way, I think we still need some “merging” of the kernels = in your case a generation of an outer wrapper kernel. Currently, the merge loopy is responsible for generating the “right sequence of calls for population of the terminal temporaries”. I insert the instruction for initialisation of the temporaries* and inlining of the tsfc before the slate operations. I think what we could do, however, is to move the second part of the kernel builder, the part that is in setup, into the merging step as well.

With moving the argument generation from gem_to_loopy (1.) *and* the things that are in the setup (2.) into the loopy merging, we would avoid all this `if builder.needs_something: “add this something as argument”` and could just do it in one go.

The only way I could *potentially* see to avoid the gluing is to move it a step up. In the end calling the TSFC kernel for Slate should be very similar to calling the TSFC kernel in PyOP2, if there is no tensor algebra happening. The conditional code structure in the setup in the kernel builder is very similar to the assembly code here https://github.com/firedrakeproject/firedrake/blob/418eaea8e307de39983cf9eb60caa8e01d266212/firedrake/assemble.py#L421 (Actually I wasted quite a lot of time on fixing that the arguments are passed into the outer kernel the same order they are passed into the ParLoop.) So the idea would be, that the Slate compilation yields the loopy kernel you depicted above and calling of the TSFC kernel/filling the tensor temporaries happens somewhere else, for example in assembly. In that case you would not be able to use slac as a standalone compiler anymore, as you do for example in the multigrid tests, because you would need provide some filling for the tensors from the outside. But maybe we could avoid code duplication.

*By generating the temporary variables for the tensor with a loopy initialiser we can drop some of those as well I think.

Reply:
>  Instead of saving the tree traversal in a list and do the translation item by item, you suggest to use the Memoizer, which does the translation recursively, right? -> Will this slow down the compilation process?

It should not. the `Memoizer` model is everywhere in the tsfc pipeline.

> But currently I am dropping the factorisation node in the slate to gem translation of the Solve. So either we would have to modify the mapping you suggested here, or we would need to deal with it somewhere else. Currently the Solve Call always does a partial piv factorisation independent of what the user might specify.

I think this is a minor issue (it's effectively deciding what callable we emit for the different things).

> Now the main point
The way you wrote slate2loopy suggests that you want to not only inline the TSFC kernel, but also the Slate kernel. You want to call Tsfc to populate some temporaries and feed this as global args to the slate loopy kernel, both inside some other wrapper kernel, I assume.

I think I hadn't thought too hard about this stage. What I think you do is:

1. get tsfc kernels for each terminal, generate a sequence of `CallInstruction`s for each of them to populate a temporary.
2. get the slate kernel (same as I do, I think)
3. prepend the instructions in the slate kernel with the `CallInstruction`s you made in step 1, and fix some things up a bit.

What I vaguely had in mind was similar but my step 3 would be "make a loopy kernel that runs the callinstructions and then the slate kernel". This "outer wrapper" would be the final kernel we send back.

I'm on the fence about which is the better approach: plausibly just updating the slate kernel with new instructions is better, but I don't know if there are gotchas in terms of having to manage name-matching everywhere (it seems like you have to work a little to modify the domains of the final loopy kernel so that the indices match up). My thought had been, given that loopy in theory has all this technology for callables, we should just use it.

> The conditional code structure in the setup in the kernel builder is very similar to the assembly code here

This is, I think, unavoidable without major rethinking of lots of things (unfortunately!)

> So the idea would be, that the Slate compilation yields the loopy kernel you depicted above and calling of the TSFC kernel/filling the tensor temporaries happens somewhere else, for example in assembly. In that case you would not be able to use slac as a standalone compiler anymore, as you do for example in the multigrid tests, because you would need provide some filling for the tensors from the outside. But maybe we could avoid code duplication.

I think getting back something from the slate compiler that isn't a "single" kernel that I can run over the mesh and get the outcome is a bad idea (as you say, we use it in various parts of the multigrid infrastructure).

Reply:
> It should not. the Memoizer model is everywhere in the tsfc pipeline.

I have just had a quick look at the Memoizer. There is definitely some recursion happening. In my understanding the point of a memoizer is to cache the result of a recursive function call, such that you can you reuse the previous result, if the same function is called on the same arguments. I assume, this does not work if you have different arguments.
I tried to understand why it is not time costly to use the Memoizer in TSFC. My thoughts are  a) the memoizer is sometimes called on functions which are actually not recursive e.g. the functions for literal rounding or replacing delta functions. b) Functions that are recursive like unrolling index sums, replacing divisions by multiplications, replacing indices in indexed nodes, and co, are only called once and their recursion stack is not very deep, or they are indeed called with the same arguments multiple times (I cannot see how the latter would be the case?)

The actual ufl2gem translation here https://github.com/firedrakeproject/tsfc/blob/b3d2a93e2d868c6e133604533d479f4e05bde1b6/tsfc/fem.py#L645 seems to use a function which does tree traversal, see https://github.com/FEniCS/ufl/blob/f042055a446f60b06440d46de449fc4f3002abc0/ufl/corealg/map_dag.py#L30, rather than recursion. The memoizer only seems to be used for gem to gem translations.

Maybe @miklos1 can comment on this? Do you think it is reasonable to use the Memoizer for the slate to gem translation?

Reply:
> I have just had a quick look at the Memoizer. There is definitely some recursion happening. In my understanding the point of a memoizer is to cache the result of a recursive function call, such that you can you reuse the previous result, if the same function is called on the same arguments. I assume, this does not work if you have different arguments.
I tried to understand why it is not time costly to use the Memoizer in TSFC. My thoughts are a) the memoizer is sometimes called on functions which are actually not recursive e.g. the functions for literal rounding or replacing delta functions. b) Functions that are recursive like unrolling index sums, replacing divisions by multiplications, replacing indices in indexed nodes, and co, are only called once and their recursion stack is not very deep, or they are indeed called with the same arguments multiple times (I cannot see how the latter would be the case?)

So the major time saving comes from treating the DAG as a DAG (not a tree) and not blowing up its perceived size. The `traversal` functions do that by maintaining a set of seen nodes and yielding only unique ones. The `Memoizer` technology achieves the same result by returning a new value for already seen nodes. So seeing the same argument multiple times comes for example if you have any common subexpressions. In both cases you want to transform that subexpression exactly once.

Both `Memoizer` and `traversal` are used in tsfc to visit the entire DAG (or bits of it). You normally use `traversal` if you want to visit the DAG and do something for each node gathering information (or perhaps collecting all nodes of a certain type). We use `Memoizer` if we want to visit the DAG and transform it, returning a new DAG but somehow preserving the structure (e.g. the UFL2GEM pass, or `simplify_abs` in ufl_utils.py).

FWIW things like `replace_delta` are recursive (the delta replacement is not recursive, but check the definition of `reuse_if_untouched` in `node.py`.

Reply:
Okay. Why use Memoizer for one and traversal for the other?

Reply:
Okay I understand, thanks @miklos1 and @wence-!